homeworks/week5/NgocTin-Le.java

```diff
@@ -0,0 +1,116 @@
+package com.dsa;
+
+//int[] boss = { 1, 3, 3, -1 };
+
+// Find lowest common boss in a tree;
+//         0  1  2   3   4  5  6  7  8
+// Input: [1, 3, 3, -1, -1, 4, 2, 8, 3]
+// 0 reports to 1, 1 reports to 3, 2 reports to 3, 3 reports to -1 (highest boss), etc.
+//      -1
+//     /  \
+//    3    4
+//  / | \   \
+// 8  2  1   5
+// |  |  |
+// 7  6  0
+// Assumption: the input array is non-empty. And  and valid inputs.
+// Requirement: No extra storage. No build-in library.
+// bossK = input[k]
+
+// 2 0
+// 1 3
+// 3 -1
+
+//7 1
+//8 3
+//3 -1
+
+// 6 0
+// 2 1
+// 3 3
+// -1
+
+// 6 5
+// 2 4
+// 3 -1
+
+import java.util.Comparator;
+import java.util.PriorityQueue;
+
+public class LowestCommonBoss {
+    public static void main(String[] args) {
+        int[] boss = {1, 3, 3, -1, -1, 4, 2, 8, 3};
+        System.out.println(lowestCommonBoss(boss, 1, 1));  // 1
+        System.out.println(lowestCommonBoss(boss, 2, 0));  // 3
+        System.out.println(lowestCommonBoss(boss, 6, 0));  // 3
+        System.out.println(lowestCommonBoss(boss, 7, 1));  // 3
+        System.out.println(lowestCommonBoss(boss, 4, 5));  // 4
+        System.out.println(lowestCommonBoss(boss, 6, 5));  // -1
+        System.out.println(lowestCommonBoss(boss, 4, 2));  // -1
+        System.out.println(lowestCommonBoss(boss, 4, 3));  // -1
+        System.out.println(lowestCommonBoss(boss, 4, -1)); // -1
+    }
+
+    // Time complexity: O(n)
```

Comment:
Assuming n is input size here.

Then optimal time complexity is O(logN) or O(H) where H = tree height.